integration/kotlinx-coroutines-firebase-android/kotlinx-coroutines-firebase-android/src/main/java/kotlinx/coroutines/experimental/firebase/android/DatabaseReference.kt

```diff
@@ -0,0 +1,266 @@
+            }
+        }
+    }
+
+    continuation.invokeOnCompletion { reference.removeEventListener(listener) }
+
+    reference.addListenerForSingleValueEvent(listener)
+}
+
+/**
+ * Coroutine to read a list of values from Firebase Database
+ *
+ * This method returns an implementation of [List] interface with the generic type informed. It's
+ * different from [kotlinx.coroutines.experimental.firebase.android.readValue] function because of
+ * the way that Firebase serialize and deserialize values.
+ *
+ * Example:
+ * ```kotlin
+ * val names = FirebaseDatabase.getInstance()
+ *     .getReference("names")
+ *     .readList(String::class.java)
+ *
+ * names.forEach(::println) // ["Adam", "Monica"]
+ * ```
+ *
+ * @param type Expected type wrapped in a Class instance.
+ * @param T The type of the expected value from Firebase Database.
+ * @return The persisted object of the type T informed.
+ */
+suspend fun <T : Any> DatabaseReference.readList(type: Class<T>): List<T> = readReferences(this, type)
+
+/**
+ * Coroutine to read a list of values from Firebase Database
+ *
+ * This method returns an implementation of [List] interface with the generic type informed. It's
+ * different from [kotlinx.coroutines.experimental.firebase.android.readValue] function because of
+ * the way that Firebase serialize and deserialize values.
+ *
+ * Example:
+ * ```kotlin
+ * val names = FirebaseDatabase.getInstance()
+ *     .getReference("names")
+ *     .readList<String>()
+ *
+ * names.forEach(::println) // ["Adam", "Monica"]
+ * ```
+ *
+ * @param T The type of the expected value from Firebase Database.
+ * @return The persisted object of the type T informed.
+ */
+suspend inline fun <reified T : Any> DatabaseReference.readList(): List<T> = readList(T::class.java)
+
+suspend fun <T : Any> DatabaseReference.saveValue(value: T): Unit =
+        setValue(value).await().let { kotlin.Unit }
+
+suspend fun <T : Any> DatabaseReference.pushValue(value: T): Unit = push().saveValue(value)
+
+private suspend fun <T : Any> awaitQuerySingleValue(query: Query, type: Class<T>): T =
+        suspendCancellableCoroutine { continuation ->
+            val listener = object : ValueEventListener {
+                override fun onDataChange(snapshot: DataSnapshot) = try {
+                    continuation.resume(snapshot.getValue(type)!!)
+                } catch (exception: Exception) {
+                    continuation.resumeWithException(exception)
+                }
+
+                override fun onCancelled(error: DatabaseError) =
+                        continuation.resumeWithException(error.toException())
+            }
+
+            query.addListenerForSingleValueEvent(listener)
+            continuation.invokeOnCompletion { query.removeEventListener(listener) }
+        }
+
+suspend fun <T : Any> Query.readValue(type: Class<T>): T = awaitQuerySingleValue(this, type)
+
+suspend inline fun <reified T : Any> Query.readValue(): T = readValue(T::class.java)
```

Comment:
Every public function should have KDoc, even if you have to copy-paste documentation a bit.
Also, it should have "redundant" `public` modifier
I'll update our contribution guide.

Reply:
> Also, it should have "redundant" `public` modifier

Changed in https://github.com/Kotlin/kotlinx.coroutines/pull/426/commits/fd3582021433e49dab9e8bd171ae7753240aa24f.